### ISSUE TITLE
js-tests application: SpiderMonkey crashes after EGL Context Lost

### ISSUE THREAD

**Issue by alonlevylll** [cocos2d/cocos2d-x#17260]
- cocos2d-js version: 3.10
- devices test on: Galaxy Tab A, LG G4 Beat 
- developing environments
   - NDK version: r10e
   - Xcode version:
   - VS version:
   - browser type and version:

After a long debug process, I noticed that it happens always after glsurface calls onSurfaceCreated on the second time, (and a new GLThread is created, I print the GLThread ID, and notice that it changes).

After it happens, the application crashes at the first call to SpiderMonkey function, in my case, at AppDelegate.cpp:

void AppDelegate::applicationWillEnterForeground() {
    ....
    // Crash HERE!
    director->getEventDispatcher()->dispatchCustomEvent("game_on_show");
}

From Android Documentation: https://developer.android.com/reference/android/opengl/GLSurfaceView.Renderer.html

> EGL Context Lost
There are situations where the EGL rendering context will be lost. This typically happens when device wakes up after going to sleep. When the EGL context is lost, all OpenGL resources (such as textures) that are associated with that context will be automatically deleted. In order to keep rendering correctly, a renderer must recreate any lost resources that it still needs. The onSurfaceCreated(GL10, EGLConfig) method is a convenient place to do this.

Steps to Reproduce:
The issue is also reproduce randomly!
But the easiest way to force the crash:
Somehow those steps forcing the GLThread to recreate:
1. Launch from "Open" button at Play Store
2. Minimize app (pause)
3. Relaunch (resume) it again from the icon

Update:
I compiled the js-tests, same issue apear there.

**Comment by minggo** [comment]
can fix it according to https://github.com/cocos2d/cocos2d-x/issues/16520

**Comment by alonlevylll** [comment]
Thanks,
However, I'm using cocos 3.10, and the generated template project already includes 

> android:configChanges="orientation|keyboardHidden|screenSize"

It doesn't help.

**Comment by dmrdev** [comment]
Have the same issue on 3.14. This issue is also present in the JSTests app and all cocos2d-x JS apps. And it has nothing to do with android:configChanges.

**Comment by dmrdev** [comment]
I also have another way to easily recreate the crash.

Launch the app using the following command:
cocos run -p android
when app is launched, press the home button in order to send it to background. Then launch the app from the launcher screen and it will crash (also it may just print "JS: too much recursion" error on the logcat and stop executing JS code in all event handlers).
Tried on Samsung Galaxy S5 (Android 6.0)

**Comment by alonlevylll** [comment]
Thanks for reopen the issue!

I did some further investigations, and found in android API:

        // Test to fix the resume background bug
        if (Build.VERSION.SDK_INT >= 11) {
            glSurfaceView.setPreserveEGLContextOnPause(true);
        }

It didn't fixed or reduced the issue also.

**Comment by dmrdev** [comment]
Little bit more info on this. SpiderMonkey stops working every time when the app activity is recreated by the system after it was destroyed.

So when we launch app using `cocos run` command, it launches the app using the following command:
`adb shell am start -n "com.example.myapp/org.cocos2dx.javascript.AppActivity"`

If this type of launch is used, Android will destroy activity immediately after the app will be moved to background, but the process will continue to exist.

But if we will start the app from the launcher (home) screen, the main activity will not be destroyed on moving to background.

**Comment by dmrdev** [comment]
Maybe it's a dirty hack but I found a workaround. Just add the attribute `android:launchMode="singleTask"` to the activity in AndroidManifest.xml, so now it is:

```
<activity android:name="org.cocos2dx.javascript.AppActivity"
            android:label="@string/app_name"
            android:launchMode="singleTask"
            android:screenOrientation="portrait"
            android:theme="@style/AppTheme"
            android:configChanges="orientation|keyboardHidden|screenSize">
      <intent-filter>
            <action android:name="android.intent.action.MAIN" />
            <category android:name="android.intent.category.LAUNCHER" />
      </intent-filter>
</activity>
```

After this Android will no longer kill the main activity and it is just resumed after process moved back from the background.

**Comment by golergka** [comment]
@dmrntpv I don't think that preventing activity from being killed can be a proper fix. Android can kill any activity in the foreground, and it should be able to restore itself from `savedInstanceState`. Debugging menu usually has "kill background activities" option to test for that behaviour.

**Comment by aaron7090** [comment]
@golergka android noob here but.. if I understand correctly, single task can still be killed and brought back?(I'll have to verify this). The only difference is that a new activity would not be created when a new intent is sent to launch it.

**Comment by netherglaive** [comment]
@dmrntpv this does not work when you go into developer settings and turn on 'do not keep activities'

**Comment by dmrdev** [comment]
@golergka, @aaron7090, I'm absolutely agree, this is only a temporary workaround while we're waiting for the proper solution from the developers.

**Comment by fuyifan** [comment]
I do recommend Cocos puts this issue as high priority. It effects all the android devices used CocosJS and resuming from background is such a common thing that will make countless abnormal behavior for all the users. I don't know whether Cocos with the lower version has the same issue. If no, I will abandon v3.14 and degrade to that version. @minggo @ricardoquesada . BTW, as i know Cocos Creator use cocos2dx on native compiled apk, don't it have the same problem?

**Comment by alonlevylll** [comment]
Someone has any ideas about that issue? @minggo @ricardoquesada

**Comment by ricardoquesada** [comment]
@fuyifan @alonlevylll No idea (and I'm no longer working on cocos2d).

**Comment by alonlevylll** [comment]
Thanks,
I think I have the full picture of that issue and I would like to know your opinion about the next steps.
So the reason is simple a new EGL thread is created (managed by Android).

- **Cocos2d-x** seems to handle this perfectly, and loading all textures again into the new EGL instance.

- **SpiderMonkey** crashes on the first call because it is compiled as single threaded, and doesn't support the new EGL thread.

So my idea would be to compile spidermonkey as JS_THREADSAFE, and wrap all calls with JS_SetContextThread and JS_ClearContextThread.

Before start test it I would like to know the reasons you not using the JS_THREADSAFE flag,
is it performance? complexity? hard to integrate? library weigh too much?

Thanks

**Comment by minggo** [comment]
As i know, JS codes are invoked in GL thread too. @pandamicro any idea?

**Comment by pandamicro** [comment]
> Before start test it I would like to know the reasons you not using the JS_THREADSAFE flag,
> is it performance? complexity? hard to integrate? library weigh too much?

It's due to performance reason, as in the design we don't want user to user spidermonkey cross thread.

> So my idea would be to compile spidermonkey as JS_THREADSAFE, and wrap all calls with JS_SetContextThread and JS_ClearContextThread.

You can try that, we recently upgraded spidermonkey to v52 in Creator's native engine which is planned for next month, but it will take longer to integrate it in Cocos2d-x

**Comment by minggo** [comment]
I uses v3 latest codes: 4be6eb532567f747ec2ad62e84893123e236e9f4 to test, what i did are:

- use cocos run -p android to build and run js-tests
- click Home button to make js-tests into background
- click js-tests app icon to run js-tests again
- no crash, but there is OpenGL ES rendering error, so can not response any input


There is not error if i did like this
- click js-tests icon to run js-tests
- click Home button to let js-tests go to background
- click js-tests icon to run js-tests again
- everything works correctly

So it seems there is an error when the app is installed and run through USB, but no crash. After the first time, there is not problem.

**Comment by kaplad** [comment]
Hello! :) Any updates @minggo?

**Comment by minggo** [comment]
No, i have not idea with this issue, i think it is something about JSB. @dumganhar any idea?

**Comment by drelaptop** [comment]
added some useful info, I think the difference between 
```
use cocos run -p android to build and run js-tests
click Home button to make js-tests into background
```
and
```
click js-tests icon to run js-tests
click Home button to let js-tests go to background
```
was caused by a android issue, refer to https://github.com/cocos2d/cocos2d-x/pull/19010#issuecomment-418249866

---
similar with @dmrdev said, we have a workaround was merged at https://github.com/cocos2d/cocos2d-x/pull/19010
> Maybe it's a dirty hack but I found a workaround. Just add the attribute android:launchMode="singleTask" to the activity in AndroidManifest.xml, so now it is:

```
            android:launchMode="singleTask"
            android:taskAffinity=""  
```
